Introduce TerminalSettingsModel project

carlos-zamora opened the pull request:
Introduces a new TerminalSettingsModel (TSM) project. This project is
responsible for (de)serializing and exposing Windows Terminal's settings
as WinRT objects.

## References
#885: TSM epic
#1564: Settings UI is dependent on this for data binding and settings access
#6904: TSM Spec

In the process of ripping out TSM from TerminalApp, a few other changes
were made to make this possible:
1. AppLogic's `ApplicationDisplayName` and `ApplicationVersion` was
   moved to `CascadiaSettings`
   - These are defined as static functions. They also no longer check if
     `AppLogic::Current()` is nullptr.
2. `enum LaunchMode` was moved from TerminalApp to TSM
3. `AzureConnectionType` and `TelnetConnectionType` were moved from the
   profile generators to their respective TerminalConnections
4. CascadiaSettings' `SettingsPath` and `DefaultSettingsPath` are
   exposed as `hstring` instead of `std::filesystem::path`
5. `Command::ExpandCommands()` was exposed via the IDL
   - This required some of the warnings to be saved to an `IVector`
     instead of `std::vector`, among some other small changes.
6. The localization resources had to be split into two halves.
   - Resource file linked in init.cpp. Verified at runtime thanks to the
     StaticResourceLoader.
7. Added constructors to some `ActionArgs`
8. Utils.h/cpp were moved to `cascadia/inc`. `JsonKey()` was moved to
   `JsonUtils`. Both TermApp and TSM need access to Utils.h/cpp.

A large amount of work includes moving to the new namespace
(`TerminalApp` --> `Microsoft::Terminal::Settings::Model`).

Fixing the tests had its own complications. Testing required us to split
up TSM into a DLL and LIB, similar to TermApp. Discussion on creating a
non-local test variant can be found in #7743.

Closes #885

carlos-zamora commented:
## Note for Reviewers
This PR is still in draft because I need to fix all of the tests. That shouldn't be too difficult. Aside from that, this PR is ready to go!

leonMSFT reviewed (commented):
I'm curious about what you said about how the "localization resources had to be split into two halves".
1. I took a look at `init.cpp` but wasn't quite sure how it's linking the resource file, could you briefly explain?
2. What if `TerminalSettingsEditor` also needed to have its own localization resources, would it have to do something similar?

DHowett commented on src/cascadia/TerminalSettingsModel/TerminalSettingsModel.vcxproj:
```diff
@@ -0,0 +1,214 @@
+<Project DefaultTargets="Build" ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <PropertyGroup Label="Globals">
+    <ProjectGuid>{CA5CAD1A-d7ec-4107-b7c6-79cb77ae2907}</ProjectGuid>
+    <ProjectName>TerminalSettingsModel</ProjectName>
```
can you change something so that this goes into a DLL called `Microsoft.Terminal.Settings.Model.dll`?

DHowett commented on src/cascadia/TerminalSettingsModel/TerminalSettingsModel.vcxproj:
```diff
@@ -0,0 +1,214 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project DefaultTargets="Build" ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <PropertyGroup Label="Globals">
+    <ProjectGuid>{CA5CAD1A-d7ec-4107-b7c6-79cb77ae2907}</ProjectGuid>
```
i hope this is unique.

carlos-zamora replied:
This used to be TerminalSettings' guid. Is that a problem?

DHowett commented on src/cascadia/TerminalSettingsModel/pch.h:
```diff
@@ -0,0 +1,54 @@
+// Including TraceLogging essentials for the binary
+#include <TraceLoggingProvider.h>
+#include <winmeta.h>
+TRACELOGGING_DECLARE_PROVIDER(g_hTerminalAppProvider);
```
rename this tracing provider.

DHowett commented on src/cascadia/TerminalSettingsModel/init.cpp:
```diff
@@ -0,0 +1,34 @@
+TRACELOGGING_DEFINE_PROVIDER(
+    g_hTerminalAppProvider,
+    "Microsoft.Windows.Terminal.App",
+    // {24a1622f-7da7-5c77-3303-d850bd1ab2ed}
```
you need to generate a new tracing guid and everything here.

DHowett replied:
the guid is based on the name. you have to use tlgguid.exe to generate it. look it up on osgwiki.

carlos-zamora replied:
I meant to ask. I know we were discussing removing some of this from TSM. Should I just remove it all? Declare and hook up the new one? Or _porque no los dos_ it and introduce the new one but leave it unused?

DHowett commented on src/cascadia/TerminalSettingsModel/Profile.cpp:
```diff
@@ -3,7 +3,7 @@
 #include "pch.h"
 #include "Profile.h"
-#include "Utils.h"
+#include "..\TerminalApp\Utils.h"
```
uh?

DHowett commented on src/cascadia/TerminalSettingsModel/KeyMapping.h:
```diff
@@ -53,22 +53,22 @@ namespace winrt::TerminalApp::implementation
         KeyMapping() = default;
 
-        TerminalApp::ActionAndArgs TryLookup(winrt::Microsoft::Terminal::TerminalControl::KeyChord const& chord) const;
+        Microsoft::Terminal::Settings::Model::ActionAndArgs TryLookup(winrt::Microsoft::Terminal::TerminalControl::KeyChord const& chord) const;
```
you should have been able to use `ActionAndArgs` directly, so long as this header file could not see the `implementation` header file for AAA.

DHowett commented on src/cascadia/TerminalSettingsModel/KeyChordSerialization.h:
```diff
@@ -0,0 +1,24 @@
+
+namespace winrt::Microsoft::Terminal::Settings::Model::factory_implementation
+{
+    struct KeyChordSerialization : KeyChordSerializationT<KeyChordSerialization, implementation::KeyChordSerialization>
```
do you need a factory for this one at all? it doesn't have any methods

DHowett replied:
if so, switch to `BASIC_FACTORY`

carlos-zamora replied:
Yeah :/ this was a surprise for me too.

DHowett commented on src/cascadia/TerminalSettingsModel/KeyChordSerialization.h:
```diff
@@ -0,0 +1,24 @@
+{
+    struct KeyChordSerialization
+    {
+        KeyChordSerialization() = default;
```
does this need to be constructable at all?

carlos-zamora replied:
Yeah. Unfortunately static runtimeclasses still need a ctor. Minorly annoyed by that haha

DHowett replied:
I don't recall this conversation coming to a close. *Did* you try it without both the factory _and_ the default ctor? There _must_ be a way to do this.,

DHowett replied:
(?)

carlos-zamora replied:
(Oh huh, sorry I missed this one.)

I tried it and that didn't let me do it. The g.cpp file is looking for the `factory_implementation` so that it can `winrt::make` it. See below:

```c++
// KeyChordSerialization.g.cpp
void* winrt_make_Microsoft_Terminal_Settings_Model_KeyChordSerialization()
{
    return winrt::detach_abi(winrt::make<winrt::Microsoft::Terminal::Settings::Model::factory_implementation::KeyChordSerialization>());
}
```

Here's the errors from the output log:
```
14>KeyMappingSerialization.cpp
14>D:\Terminal\src\cascadia\TerminalSettingsModel\Generated Files\KeyChordSerialization.g.cpp(5,111): error C2039: 'KeyChordSerialization': is not a member of 'winrt::Microsoft::Terminal::Settings::Model::factory_implementation' (compiling source file KeyChordSerialization.cpp)
14>D:\Terminal\src\cascadia\TerminalSettingsModel\KeyChordSerialization.h(20): message : see declaration of 'winrt::Microsoft::Terminal::Settings::Model::factory_implementation' (compiling source file KeyChordSerialization.cpp)
14>D:\Terminal\src\cascadia\TerminalSettingsModel\Generated Files\KeyChordSerialization.g.cpp(5,132): error C2065: 'KeyChordSerialization': undeclared identifier (compiling source file KeyChordSerialization.cpp)
14>D:\Terminal\src\cascadia\TerminalSettingsModel\Generated Files\KeyChordSerialization.g.cpp(5,37): error C2672: 'winrt::make': no matching overloaded function found (compiling source file KeyChordSerialization.cpp)
14>D:\Terminal\src\cascadia\TerminalSettingsModel\Generated Files\KeyChordSerialization.g.cpp(5,134): error C2974: 'winrt::make': invalid template argument for 'D', type expected (compiling source file KeyChordSerialization.cpp)
14>D:\Terminal\src\cascadia\TerminalSettingsModel\Generated Files\winrt\base.h(7401): message : see declaration of 'winrt::make' (compiling source file KeyChordSerialization.cpp)
14>PowershellCoreProfileGenerator.cpp
```

DHowett commented on src/cascadia/TerminalSettingsModel/Command.idl:
```diff
@@ -0,0 +1,30 @@
+        Boolean HasNestedCommands { get; };
+        Windows.Foundation.Collections.IMapView<String, Command> NestedCommands { get; };
+
+        static void ExpandCommands(Windows.Foundation.Collections.IMap<String, Command> commands,
```
should this be a MapView?

carlos-zamora replied:
Nope. We're actively removing and inserting commands.

DHowett commented on src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp:
```diff
@@ -834,7 +832,7 @@ std::optional<std::string> CascadiaSettings::_ReadUserSettings()
         // migrate any file we find. We're using MoveFile in case their settings.json
         // is a symbolic link.
-        auto pathToLegacySettingsFile{ pathToSettingsFile };
+        std::filesystem::path pathToLegacySettingsFile{ pathToSettingsFile.c_str() };
```
```suggestion
        std::filesystem::path pathToLegacySettingsFile{ std::wstring_view{ pathToSettingsFile } };
```

if you find yourself using `c_str`, consider alternatives.

DHowett commented on src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp:
```diff
@@ -484,7 +482,7 @@ bool CascadiaSettings::_AppendDynamicProfilesToUserSettings()
             if (profileJson.isObject())
             {
-                const auto profileImpl = winrt::get_self<winrt::TerminalApp::implementation::Profile>(profile);
+                const auto profileImpl = winrt::get_self<winrt::Microsoft::Terminal::Settings::Model::implementation::Profile>(profile);
```
there must be a shorter way to write this

DHowett commented on src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp:
```diff
@@ -5,7 +5,7 @@
 #include "CascadiaSettings.h"
 #include "../../types/inc/utils.hpp"
-#include "utils.h"
+#include "..\TerminalApp\Utils.h"
```
hey come now, follow the slash format that's already in use in the file

DHowett commented on src/cascadia/TerminalSettingsModel/ActionArgs.h:
```diff
@@ -308,9 +314,12 @@ namespace winrt::TerminalApp::implementation
-        GETSET_PROPERTY(winrt::TerminalApp::SplitState, SplitStyle, winrt::TerminalApp::SplitState::Automatic);
-        GETSET_PROPERTY(winrt::TerminalApp::NewTerminalArgs, TerminalArgs, nullptr);
-        GETSET_PROPERTY(winrt::TerminalApp::SplitType, SplitMode, winrt::TerminalApp::SplitType::Manual);
+        SplitPaneArgs(SplitState style, Model::NewTerminalArgs terminalArgs) :
```
be sure to take winrt class arguments by const reference to avoid additoinal reference counting overhead

DHowett replied:
i wonder if the compiler can see through this and turn it into a `move`...

DHowett commented on src/cascadia/TerminalApp/Tab.cpp:
```diff
@@ -1043,15 +1042,11 @@ namespace winrt::TerminalApp::implementation
-        auto args = winrt::make_self<implementation::SwitchToTabArgs>();
-        args->TabIndex(_TabViewIndex);
+        SwitchToTabArgs args{ _TabViewIndex };
+        ActionAndArgs focusTabAction{ShortcutAction::SwitchToTab, args};
```
format didn't bark?

DHowett commented on src/cascadia/TerminalApp/AppLogic.cpp:
```diff
@@ -718,7 +719,7 @@ namespace winrt::TerminalApp::implementation
     {
         // Get the containing folder.
-        const auto settingsPath{ CascadiaSettings::GetSettingsPath() };
+        const std::filesystem::path settingsPath{ CascadiaSettings::SettingsPath().c_str() };
```
try casting through wstring_view if possible.

DHowett commented on src/cascadia/TerminalApp/AppCommandlineArgs.cpp:
```diff
@@ -319,20 +315,19 @@ void AppCommandlineArgs::_buildFocusTabParser()
-                focusTabAction->Action(_focusNextTab ? ShortcutAction::NextTab : ShortcutAction::PrevTab);
-                _startupActions.push_back(*focusTabAction);
+                focusTabAction.Action(_focusNextTab ? ShortcutAction::NextTab : ShortcutAction::PrevTab);
+                _startupActions.push_back(focusTabAction);
```
you can `emplace_back(std::move(focusTabAction))` now to save a refcount

DHowett reviewed (changes requested):
96/102

carlos-zamora commented on src/cascadia/TerminalApp/TerminalPage.cpp:
```diff
@@ -330,7 +327,7 @@ namespace winrt::TerminalApp::implementation
                 else
                 {
-                    return;
+                    co_return;
```
Michael made this same change in his DefTerm implementation so I'm not sure why it appears as an issue sometimes. Either way, I think this is more correct because we're `co_await`-ing right before it. So the idea is you're supposed to `co_return` on a `co_await` or `fire_and_forget`.

DHowett commented on src/cascadia/TerminalApp/TerminalAppLib.vcxproj:
```diff
@@ -160,7 +160,7 @@
     <ClCompile Include="ColorHelper.cpp" />
     <ClCompile Include="DebugTapConnection.cpp" />
-    <ClCompile Include="Utils.cpp" />
+    <ClCompile Include="$(OpenConsoleDir)\src\cascadia\inc\Utils.cpp" />
```
_hold up, absolutely not_

do not include some other random folder's cpp files

DHowett replied:
**inc is for headers only**

zadjii-msft replied:
It looks like there's only one function that's actually implemented in Utils.cpp, we could _probably_ just move that to the header, and avoid this issue entirely

zadjii-msft reviewed (commented):
Okay, I'm done with all 133 in this PR, and my only concern is over the same `utils.cpp` thing Dustin brought up

DHowett reviewed (commented):
hitting submit b/c queued comments?

carlos-zamora commented:
Changes for the next commit(s):
- [ ] Try to kill `JsonKey`
- [ ] Move `VisualizeControlCodes` into TIL
- [ ] Refactor `Command` to expose icon like Profile does (loading image happens outside of settings model)

carlos-zamora commented:
> * [x]  Try to kill `JsonKey` (or move to `JsonUtils`)

Oh...I already had it in `JsonUtils`.

> * [x]  Move `VisualizeControlCodes` into TIL

Moved and added a test. Had to remove a `noexcept` because of audit mode.

> * [x]  Refactor `Command` to expose icon like Profile does (loading image happens outside of settings model)

CommandPalette data binds itself to the IconSource of `Command`. Refactoring such that that isn't that case felt a bit more risky than I'd like it to be for this PR. I still moved `RefreshIcon()` from `Command` into `TerminalPage`, which should be good enough tbh. I'll open a follow-up item if you feel strongly about this though.

zadjii-msft commented on src/cascadia/TerminalApp/TerminalPage.cpp:
```diff
@@ -80,6 +80,70 @@ namespace winrt::TerminalApp::implementation
+    // - <none>
+    // Return Value:
+    // - <none>
+    static void _refreshIcon(const Command& cmd)
```
Wow I can't believe we let me ship this in the first place. Really, Command should just have an internal-only `hstring _iconPath`, and `Command::IconSource()` should probably do this dance to create the icon source. That's not really on you though, you're working with what I gave you, so I'm gonna sign off w/o this. Presumably this is what #7784 is talking about

carlos-zamora replied:
That's exactly what #7784 is haha. I'll start working on that. Dustin gave me a rundown of how it can work with converters.

DHowett commented on src/cascadia/LocalTests_SettingsModel/SerializationTests.cpp:
```diff
@@ -0,0 +1,2373 @@
+    // an updated TAEF that will let us install framework packages when the test
+    // package is deployed. Until then, these tests won't deploy in CI.
+
+    class SerializationTests : public JsonTestClass
```
these are Deserialization tests ...

carlos-zamora replied:
Fixed. Bear with me. This is one of those small things that I'll be hitting _a lot_ this month.

DHowett commented on src/cascadia/LocalTests_TerminalApp/TabTests.cpp:
```diff
@@ -276,12 +275,8 @@ namespace TerminalAppLocalTests
 
-        VerifyParseSucceeded(settingsJson0);
-        auto settings0 = winrt::make_self<implementation::CascadiaSettings>(false);
+        CascadiaSettings settings0{ winrt::hstring{ til::u8u16(settingsJson0) } };
```
weird: this one grew an `hstring` conversion but the other two didn't need it ...

DHowett replied:
(the ones lower in the file)

DHowett commented on src/cascadia/TerminalSettingsModel/Profile.cpp:
```diff
@@ -3,7 +3,7 @@
 #include "pch.h"
 #include "Profile.h"
-#include "Utils.h"
+#include "../inc/Utils.h"
```
sorry, _what_ lives in inc/utils.h now? i couldn't tell by searching the PR :) and I can;t find the file in your branch

how does this compile o_O

carlos-zamora replied:
Huh. This was pointing to "WinRTUtils/inc/Utils". Not needed anymore.

DHowett commented on src/cascadia/TerminalApp/TerminalPage.cpp:
```diff
@@ -83,7 +80,71 @@ namespace winrt::TerminalApp::implementation
 
-    static void _recursiveUpdateCommandIcons(IMapView<winrt::hstring, winrt::TerminalApp::Command> commands)
+    // Method Description:
+    // - Actually initialize the IconSource. Supports a variety of icons:
```
`the`? this comment seems like it could use some rework

DHowett commented on src/cascadia/TerminalSettingsModel/ColorScheme.h:
```diff
@@ -55,12 +55,12 @@ namespace winrt::TerminalApp::implementation
-namespace winrt::TerminalApp::factory_implementation
+namespace winrt::Microsoft::Terminal::Settings::Model::factory_implementation
 {
     BASIC_FACTORY(ColorScheme);
```
whoa you can construct a colorscheme out of the idl?

carlos-zamora replied:
Yeah. We don't need it right now, but we'll need it for Settings UI soon. Removed.

DHowett commented on src/cascadia/TerminalSettingsModel/dll/Microsoft.Terminal.Settings.Model.vcxproj:
```diff
@@ -0,0 +1,108 @@
+      <AdditionalOptions Condition="'$(Platform)'=='Win32'">/INCLUDE:_DllMain@12</AdditionalOptions>
+      <AdditionalOptions Condition="'$(Platform)'!='Win32'">/INCLUDE:DllMain</AdditionalOptions>
+    </Link>
+    <Reference>
```
```suggestion
    <Reference>
        <!-- Do not propagate microsoft.ui.xaml upwards as a private dependency. -->
```

comment weird hacks.

DHowett commented on src/cascadia/TerminalSettingsModel/dll/pch.h:
```diff
@@ -0,0 +1,12 @@
+// Header for platform projection include files
+//
+
+#pragma once
+
+// This file can be empty - the pch.h in TerminalSettingsModel/lib does the heavy lifting
+// of including all the headers we need. As this project is just a dll wrapper,
```
are these craps still necessary

DHowett replied:
did we copy it because we copied it, or did we copy it because it will break otherwise

carlos-zamora replied:
Copied because it was copied. Just removed it. Wasn't needed at all.

carlos-zamora commented:
Urgh. _Technically_, y'all should be reading/merging https://github.com/microsoft/terminal/pull/6904 first. So go read that * *grumble* *

DHowett commented on src/cascadia/TerminalApp/DebugTapConnection.cpp:
```diff
@@ -3,7 +3,7 @@
 #include "pch.h"
 #include "DebugTapConnection.h"
-#include "Utils.h"
+#include "../inc/Utils.h"
```
........................................

DHowett commented on src/cascadia/TerminalSettingsModel/GlobalAppSettings.cpp:
```diff
@@ -5,14 +5,14 @@
 #include "../../types/inc/Utils.hpp"
 #include "../../inc/DefaultSettings.h"
-#include "Utils.h"
+#include "../inc/Utils.h"
```
?????????????

DHowett commented on src/cascadia/TerminalSettingsModel/KeyMappingSerialization.cpp:
```diff
@@ -10,11 +10,11 @@
 #include "ActionAndArgs.h"
 #include "KeyChordSerialization.h"
-#include "Utils.h"
+#include "../inc/Utils.h"
```
!!!!!!!!!!!!!!!!!!!!!!!!!!

DHowett commented on src/cascadia/TerminalSettingsModel/CascadiaSettings.cpp:
```diff
@@ -100,7 +108,7 @@ IObservableVector<winrt::TerminalApp::Profile> CascadiaSettings::Profiles() cons
-winrt::TerminalApp::KeyMapping CascadiaSettings::KeyMap() const noexcept
+winrt::Microsoft::Terminal::Settings::Model::KeyMapping CascadiaSettings::KeyMap() const noexcept
 {
     return _globals->KeyMap();
```
>als->KeyMap(); [](start = 16, length = 14)

Testing with codeflow
